include/deal.II/distributed/fully_distributed_tria.h

```diff
@@ -0,0 +1,406 @@
+      class ParallelDistributed;
+    } // namespace Policy
+  }   // namespace DoFHandlerImplementation
+} // namespace internal
+
+namespace FETools
+{
+  namespace internal
+  {
+    template <int, int, class>
+    class ExtrapolateImplementation;
+  }
+} // namespace FETools
+
+// forward declaration of the data type for periodic face pairs
+namespace GridTools
+{
+  template <typename CellIterator>
+  struct PeriodicFacePair;
+}
+#endif
+
+namespace parallel
+{
+  /**
+   * A namespace for the fully distributed Triangulation.
+   *
+   * @ingroup parallel
+   */
+  namespace fullydistributed
+  {
+    /**
+     * Information needed for each locally relevant cell, stored in
+     * ConstructionData and used during construction of a
+     * parallel::fullydistributed::Triangulation. This struct stores
+     * the cell id, the subdomain_id and the level_subdomain_id as well as
+     * information related to manifold_id and boundary_id.
+     *
+     * @author Peter Munch, 2019
+     */
+    template <int dim>
+    struct CellInfo
+    {
+      /**
+       * Constructor.
+       */
+      CellInfo() = default;
+
+      /**
+       * Unique CellID of the cell.
+       */
+      CellId::binary_type id;
+
+      /**
+       * subdomain_id of the cell.
+       */
+      types::subdomain_id subdomain_id;
+
+      /**
+       * level_subdomain_id of the cell.
+       */
+      types::subdomain_id level_subdomain_id;
+
+      /**
+       * Manifold id of the cell.
+       */
+      types::material_id manifold_id;
+
+      /**
+       * Manifold id of all lines of the cell.
+       *
+       * @note Only used for 2D and 3D.
+       */
+      std::array<types::material_id, GeometryInfo<dim>::lines_per_cell>
+        manifold_line_ids;
+
+      /**
+       * Manifold id of all faces of the cell.
+       *
+       * @note Only used for 3D.
+       */
+      std::array<types::material_id, GeometryInfo<dim>::quads_per_cell>
+        manifold_quad_ids;
+
+      /**
+       * List of face number and boundary id of all non-internal faces of the
+       * cell.
+       */
+      std::vector<std::pair<unsigned int, types::boundary_id>> boundary_ids;
+    };
+
+
+
+    /**
+     * Data used to construct a parallel::fullydistributed::Triangulation.
+     *
+     * @author Peter Munch, 2019
+     */
+    template <int dim, int spacedim>
+    struct ConstructionData
+    {
+      /**
+       * Cells of the locally-relevant coarse-grid triangulation.
+       */
+      std::vector<CellData<dim>> coarse_cells;
+
+      /**
+       * Vertices of the locally-relevant coarse-grid triangulation.
+       */
+      std::vector<Point<spacedim>> coarse_cell_vertices;
+
+      /**
+       * Mapping from coarse-grid index to coarse-grid id.
+       */
+      std::vector<types::coarse_cell_id> coarse_cell_index_to_coarse_cell_id;
+
+      /**
+       * CellInfo for each locally relevant cell on each level.
+       */
+      std::vector<std::vector<CellInfo<dim>>> cell_infos;
```

Comment:
I missed at first reading that this is a vector of vectors. You'll have to explain what the outer and inner level mean.

I did read up on this in the implementation, and I have to admit that I don't like it particularly because you have to repeat some much information on the refined levels that can be computed from the coarser parent cells. As I'll point out below, you also have a quadratic loop. I'd much rather you find a different way to describe *how* the coarse mesh should be refined, rather than explicitly listing the details of the cells you want.